Add params collections

BillWagner opened the pull request:
Update descriptions for the `params` modifier to allow collections.

Contributes to #40005

This might be easiest to review commit by commit:

1. **Update What's new, language reference, and programming guide where `params` is mentioned.** Search the content for `params` and make updates as needed.
1. **Review articles and samples with smaller scope changes.** A general edit pass.
1. **Review methods.md**: Major edit pass, move samples and trim unused samples. Small updates to the sample code.
1. **Review programming-guide Object Initializers article**: Major edit pass, move samples. Small updates to the samples.
1. **Fix warnings from related samples moving**.
1. **Bug fix**:  Fixes #40796

<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/language-reference/keywords/method-parameters.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/language-reference/keywords/method-parameters.md) | [Method Parameters](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/method-parameters?branch=pr-en-us-40978) |
| [docs/csharp/language-reference/operators/lambda-expressions.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/language-reference/operators/lambda-expressions.md) | ["Lambda expressions - Lambda expressions and anonymous functions"](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/lambda-expressions?branch=pr-en-us-40978) |
| [docs/csharp/methods.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/methods.md) | [Methods in C\#](https://review.learn.microsoft.com/en-us/dotnet/csharp/methods?branch=pr-en-us-40978) |
| [docs/csharp/misc/cs0225.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/misc/cs0225.md) | [Compiler Error CS0225](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/cs0225?branch=pr-en-us-40978) |
| [docs/csharp/programming-guide/classes-and-structs/how-to-initialize-a-dictionary-with-a-collection-initializer.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/programming-guide/classes-and-structs/how-to-initialize-a-dictionary-with-a-collection-initializer.md) | [How to initialize a dictionary with a collection initializer (C# Programming Guide)](https://review.learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/how-to-initialize-a-dictionary-with-a-collection-initializer?branch=pr-en-us-40978) |
| [docs/csharp/programming-guide/classes-and-structs/how-to-initialize-objects-by-using-an-object-initializer.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/programming-guide/classes-and-structs/how-to-initialize-objects-by-using-an-object-initializer.md) | [docs/csharp/programming-guide/classes-and-structs/how-to-initialize-objects-by-using-an-object-initializer](https://review.learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/how-to-initialize-objects-by-using-an-object-initializer?branch=pr-en-us-40978) |
| [docs/csharp/programming-guide/classes-and-structs/object-and-collection-initializers.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/programming-guide/classes-and-structs/object-and-collection-initializers.md) | [Object and Collection Initializers (C# Programming Guide)](https://review.learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/object-and-collection-initializers?branch=pr-en-us-40978) |
| [docs/csharp/whats-new/csharp-13.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/whats-new/csharp-13.md) | [What's new in C# 13](https://review.learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-13?branch=pr-en-us-40978) |


<!-- PREVIEW-TABLE-END -->

BillWagner committed:
Add params collections

Update descriptions for the `params` modifier to allow collections.

Fixes #40005

BillWagner committed:
edit pass, phase one.

Update sample code and markdown extensions.

Grammar pass.

BillWagner committed:
edit pass

Move samples, edit for grammar and update.

BillWagner committed:
edit pass

Edit pass on the final article.

BillWagner committed:
Fix moved snippet links

BillWagner committed:
Modify Sample

This commit fixes #40796

gewarren commented on docs/csharp/methods.md:
```diff
@@ -32,112 +29,100 @@ These parts together form the method signature.
+:::code language="csharp" source="snippets/methods/methods40.cs" id="snippet40":::
 
-<a name="invocation"></a>
+The `Motorcycle` class includes an overloaded method, `Drive`. Two methods have the same name, but you must differentiate them by their parameter types.
```
```suggestion
The `Motorcycle` class includes an overloaded method, `Drive`. Two methods have the same name, but are differentiated by their parameter types.
```

gewarren replied:
It's describing someone else's code, not code the reader would/could write.

gewarren commented on docs/csharp/methods.md:
```diff
@@ -32,112 +29,100 @@ These parts together form the method signature.
-Sometimes, the requirement that you specify the exact number of arguments to your method is restrictive. By using the `params` keyword to indicate that a parameter is a parameter array, you allow your method to be called with a variable number of arguments. The parameter tagged with the `params` keyword must be an array type, and it must be the last parameter in the method's parameter list.
+Sometimes, the requirement that you specify the exact number of arguments to your method is restrictive. By using the `params` keyword to indicate that a parameter is a parameter collection, you allow your method to be called with a variable number of arguments. The parameter tagged with the `params` keyword must be a collection type, and it must be the last parameter in the method's parameter li ...
 
 A caller can then invoke the method in either of four ways:
```
```suggestion
A caller can then invoke the method in either of four ways for the `params` parameter:
```

gewarren commented on docs/csharp/misc/cs0225.md:
```diff
@@ -10,9 +10,9 @@ ms.assetid: 0b0cd72b-c47a-44d1-9b27-d1a1fad06807
+The params parameter must be a single-dimensional array or collection
   
- When using the [params](../language-reference/keywords/method-parameters.md#params-modifier) keyword, you must specify a single-dimensional array of the data type. For more information, see [Methods](../programming-guide/classes-and-structs/methods.md).  
+ When using the [params](../language-reference/keywords/method-parameters.md#params-modifier) keyword, you must specify a single-dimensional or collection of the data type. For more information, see [Methods](../programming-guide/classes-and-structs/methods.md).  
```
```suggestion
 When using the [params](../language-reference/keywords/method-parameters.md#params-modifier) keyword, you must specify a single-dimensional array or collection of the data type. For more information, see [Methods](../programming-guide/classes-and-structs/methods.md).  
```

gewarren commented on docs/csharp/programming-guide/classes-and-structs/object-and-collection-initializers.md:
```diff
@@ -125,48 +124,48 @@ Required init-only properties support immutable structures while allowing natura
 ## Object Initializers with collection read-only property initialization
 
-Some classes may have collection properties where the property is read-only, like the `Cats` property of `CatOwner` in the following case:
+Some classes can have collection properties where the property is read-only, like the `Cats` property of `CatOwner` in the following case:
```
```suggestion
Some classes might have collection properties where the property is read-only, like the `Cats` property of `CatOwner` in the following case:
```

gewarren commented on docs/csharp/whats-new/csharp-13.md:
```diff
@@ -21,11 +22,17 @@ New features are added to the "What's new in C#" page when they're available in
 
+## `params` collections
+
+The `params` modifier isn't limited to array types. You can now use `param` with any recognized collection type. In addition to arrays, <xref:System.Span%601?displayProperty=nameWithType>, <xref:System.ReadOnlySpan%601?displayProperty=nameWithType>, types that implement <xref:System.Collections.Generic.IEnumerable%601?displayProperty=nameWithType> in addition to having an `Add` method. In additio ...
```
```suggestion
The `params` modifier isn't limited to array types. You can now use `params` with any recognized collection type, including <xref:System.Span%601?displayProperty=nameWithType>, <xref:System.ReadOnlySpan%601?displayProperty=nameWithType>, and types that implement <xref:System.Collections.Generic.IEnumerable%601?displayProperty=nameWithType> and have an `Add` method. In addition to concrete types, the interfaces <xref:System.Collections.Generic.IEnumerable%601?displayProperty=nameWithType>, <xref:System.Collections.Generic.IReadOnlyCollection%601?displayProperty=nameWithType>, <xref:System.Collections.Generic.IReadOnlyList%601?displayProperty=nameWithType>, <xref:System.Collections.Generic.ICollection%601?displayProperty=nameWithType>, and <xref:System.Collections.Generic.IList%601?displayProperty=nameWithType> can also be used.
```

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Genevieve Warren <24882762+gewarren@users.noreply.github.com>

BillWagner committed:
Clarify the create method

Respond to feedback. The Create method must follow the same pattern used for collection expressions.

BillWagner commented on docs/csharp/language-reference/keywords/method-parameters.md:
```diff
@@ -197,16 +197,39 @@ The only method call where the argument is passed by reference is the final one.
+  - `System.Span<T>`
+  - `System.ReadOnlySpan<T>`  
+  in which cases the *element type* is `T`
+- A *type* with an accessible *create method* with a corresponding *element type*.
```
It must follow the same pattern used in collection expressions. I  added a clarifying sentence and a link to the collection builder article.

BillWagner committed:
warning
